### PR TITLE
*: rollback schema in the tracker; fix save table checkpoint in optimistic mode

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

fix https://github.com/pingcap/dm/issues/585.

### What is changed and how it works?

rollback tracked schema when pausing/stopping.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has persistent data change

Side effects

 - Increased code complexity

### CODE REVIEW

File: syncer/checkpoint.go [hunk 1]
```diff
@@ -530,15 +545,17 @@ func (cp *RemoteCheckPoint) CheckGlobalPoint() bool {
...
-	cp.globalPoint.rollback()
+	cp.globalPoint.rollback(schemaTracker, "")
 	for schema, mSchema := range cp.points {
 		for table, point := range mSchema {
 			logger := cp.logCtx.L().WithFields(zap.String("schema", schema), zap.String("table", table))
-			logger.Info("rollback checkpoint", log.WrapStringerField("checkpoint", point))
-			if point.rollback() {
+			logger.Debug("try to rollback checkpoint", log.WrapStringerField("checkpoint", point))
+			from := point.MySQLLocation()
+			if point.rollback(schemaTracker, schema) {
```
Comment (csuzhangxc) [inline syncer/checkpoint.go:554]: also see https://github.com/pingcap/dm/issues/629.

File: syncer/checkpoint.go [hunk 2]
```diff
@@ -96,11 +96,25 @@ func (b *binlogPoint) flush() {
 	b.flushedTI = b.ti
 }
 
-func (b *binlogPoint) rollback() (isSchemaChanged bool) {
+func (b *binlogPoint) rollback(schemaTracker *schema.Tracker, schema string) (isSchemaChanged bool) {
 	b.Lock()
 	defer b.Unlock()
-	b.location = b.flushedLocation
-	if isSchemaChanged = b.ti != b.flushedTI; isSchemaChanged {
+	b.location = b.flushedLocation.Clone()
```
Comment (csuzhangxc) [inline syncer/checkpoint.go:102]: also `.Clone`.

**Comment by csuzhangxc** [comment]
@WangXiangUSTC @lichunzhu PTAL

File: syncer/optimist.go [hunk 3]
```diff
@@ -92,10 +96,8 @@ func (s *Syncer) handleQueryEventOptimistic(
 		return err
 	}
 
-	s.tctx.L().Info("save table checkpoint", zap.String("event", "query"),
-		zap.String("schema", upSchema), zap.String("table", upTable),
-		zap.Strings("ddls", needHandleDDLs), log.WrapStringerField("location", ec.currentLocation))
-	s.saveTablePoint(upSchema, upTable, ec.currentLocation.Clone())
```
Comment (csuzhangxc) [inline syncer/optimist.go:98]: found a bug relative to https://github.com/pingcap/dm/issues/584.

File: syncer/test_injector.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+// handleFlushCheckpointStage handles failpoint of `FlushCheckpointStage`.
+func handleFlushCheckpointStage(expectStage, maxStage int, stageStr string) error {
+	if testInjector.flushCheckpointStage != expectStage {
+		return nil
+	}
+
+	log.L().Info("set FlushCheckpointStage", zap.String("failpoint", "FlushCheckpointStage"), zap.Int("stage", testInjector.flushCheckpointStage))
+	if testInjector.flushCheckpointStage == maxStage {
+		testInjector.flushCheckpointStage = -1 // disable for following stages.
```
Comment (lichunzhu) [inline syncer/test_injector.go:42]: After set `flushCheckpointStage` to -1, will `flushCheckpointStage` increase to `maxStage` later?

Comment (csuzhangxc) [reply syncer/test_injector.go:42]: No, So how about setting it to MaxInt?

Comment (lichunzhu) [reply syncer/test_injector.go:42]: I understand it. I think -1 is just OK.

File: syncer/checkpoint.go [hunk 5]
```diff
@@ -96,11 +96,25 @@ func (b *binlogPoint) flush() {
...
+	// and after a new DDL applied to the schema, the returned pointer of `model.TableInfo` changed now.
+	trackedTi, _ := schemaTracker.GetTable(schema, b.ti.Name.O) // ignore the returned error, only compare `trackerTi` is enough.
+	// may three versions of schema exist:
+	// - the one tracked in the TiDB-with-mockTiKV.
+	// - the one in the checkpoint but not flushed.
+	// - the one in the checkpoint and flushed.
+	// if any of them are not equal, then we rollback them:
+	// - set the one in the checkpoint but not flushed to the one flushed.
+	// - set the one tracked to the one in the checkpoint by the caller of this method (both flushed and not flushed are the same now)
+	if isSchemaChanged = trackedTi != b.ti || b.ti != b.flushedTI; isSchemaChanged {
```
Comment (WangXiangUSTC) [inline syncer/checkpoint.go:117]: how about adding some `()`

Comment (csuzhangxc) [reply syncer/checkpoint.go]: added in https://github.com/pingcap/dm/pull/625/commits/846c6b5ccf169eb67f81cf50eb190dba984e40c2.

**Pull request merged**